### PR TITLE
feat: a simp lemma for finprod over propositions

### PR DESCRIPTION
**Pull request by alexjbest**
This is useful when a nested finprod includes a prop variable

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/BigOperators/Finprod.lean [hunk 1]
```diff
@@ -351,6 +351,23 @@ theorem finprod_inv_distrib [DivisionCommMonoid G] (f : α → G) : (∏ᶠ x, (
 #align finprod_inv_distrib finprod_inv_distrib
 #align finsum_neg_distrib finsum_neg_distrib
 
+@[to_additive (attr := simp)]
+theorem finprod_prop (P : Prop) [Decidable P] (b : M) : ∏ᶠ _x : P, b = if P then b else 1 := by
+  rw [finprod, dif_pos (toFinite _), Finset.prod_const, mulSupport]
+  simp only [comp_apply, ne_eq]
+  by_cases hb : b = 1
+  . simp [hb]
```
Comment (alexjbest) [inline Mathlib/Algebra/BigOperators/Finprod.lean:359]:
```suggestion
    · simp [hb]
```

File: Mathlib/Algebra/BigOperators/Finprod.lean [hunk 2]
```diff
@@ -351,6 +351,23 @@ theorem finprod_inv_distrib [DivisionCommMonoid G] (f : α → G) : (∏ᶠ x, (
...
+  . simp only [hb, not_false_eq_true, setOf_true, toFinite_toFinset, toFinset_univ]
```
Comment (alexjbest) [inline Mathlib/Algebra/BigOperators/Finprod.lean:360]:
```suggestion
    · simp only [hb, not_false_eq_true, setOf_true, toFinite_toFinset, toFinset_univ]
```

File: Mathlib/Algebra/BigOperators/Finprod.lean [hunk 3]
```diff
@@ -351,6 +351,23 @@ theorem finprod_inv_distrib [DivisionCommMonoid G] (f : α → G) : (∏ᶠ x, (
...
+    split_ifs with h
+    . have : Unique (P) := uniqueProp h
```
Comment (alexjbest) [inline Mathlib/Algebra/BigOperators/Finprod.lean:362]:
```suggestion
      · have : Unique (P) := uniqueProp h
```

File: Mathlib/Algebra/BigOperators/Finprod.lean [hunk 4]
```diff
@@ -351,6 +351,23 @@ theorem finprod_inv_distrib [DivisionCommMonoid G] (f : α → G) : (∏ᶠ x, (
...
+      have : Unique (PLift P)
+      infer_instance
+      simp
+    . have : IsEmpty P := ⟨h⟩
```
Comment (alexjbest) [inline Mathlib/Algebra/BigOperators/Finprod.lean:366]:
```suggestion
      · have : IsEmpty P := ⟨h⟩
```

File: Mathlib/Algebra/BigOperators/Finprod.lean [hunk 5]
```diff
@@ -351,6 +351,23 @@ theorem finprod_inv_distrib [DivisionCommMonoid G] (f : α → G) : (∏ᶠ x, (
...
 
+@[to_additive (attr := simp)]
+theorem finprod_prop (P : Prop) [Decidable P] (b : M) : ∏ᶠ _x : P, b = if P then b else 1 := by
+  rw [finprod, dif_pos (toFinite _), Finset.prod_const, mulSupport]
+  simp only [comp_apply, ne_eq]
+  by_cases hb : b = 1
+    · simp [hb]
+    · simp only [hb, not_false_eq_true, setOf_true, toFinite_toFinset, toFinset_univ]
+    split_ifs with h
+      · have : Unique (P) := uniqueProp h
```
Comment (alexjbest) [inline Mathlib/Algebra/BigOperators/Finprod.lean:362]:
```suggestion
    · have : Unique (P) := uniqueProp h
```

File: Mathlib/Algebra/BigOperators/Finprod.lean [hunk 6]
```diff
@@ -351,6 +351,23 @@ theorem finprod_inv_distrib [DivisionCommMonoid G] (f : α → G) : (∏ᶠ x, (
...
+      have : Unique (PLift P)
+      infer_instance
+      simp
+      · have : IsEmpty P := ⟨h⟩
```
Comment (alexjbest) [inline Mathlib/Algebra/BigOperators/Finprod.lean:366]:
```suggestion
    · have : IsEmpty P := ⟨h⟩
```

File: Mathlib/Algebra/BigOperators/Finprod.lean [hunk 7]
```diff
@@ -351,6 +351,19 @@ theorem finprod_inv_distrib [DivisionCommMonoid G] (f : α → G) : (∏ᶠ x, (
 #align finprod_inv_distrib finprod_inv_distrib
 #align finsum_neg_distrib finsum_neg_distrib
 
+@[to_additive (attr := simp)]
+theorem finprod_prop (P : Prop) [Decidable P] (b : M) : ∏ᶠ _x : P, b = if P then b else 1 := by
+  rw [finprod, dif_pos (toFinite _), Finset.prod_const, mulSupport]
+  simp only [comp_apply, ne_eq]
+  by_cases hb : b = 1
```
Comment (ocfnash) [inline Mathlib/Algebra/BigOperators/Finprod.lean:358]:
```suggestion
  rcases eq_or_ne with rfl | hb; simp
```
(untested)

**Review by ocfnash — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/Algebra/BigOperators/Finprod.lean [hunk 8]
```diff
@@ -351,6 +351,19 @@ theorem finprod_inv_distrib [DivisionCommMonoid G] (f : α → G) : (∏ᶠ x, (
 #align finprod_inv_distrib finprod_inv_distrib
 #align finsum_neg_distrib finsum_neg_distrib
 
+@[to_additive (attr := simp)]
+theorem finprod_prop (P : Prop) [Decidable P] (b : M) : ∏ᶠ _x : P, b = if P then b else 1 := by
+  rw [finprod, dif_pos (toFinite _), Finset.prod_const, mulSupport]
+  simp only [comp_apply, ne_eq]
+  rcases eq_or_ne with rfl | hb; simp
+  · simp [hb]
```
Comment (alexjbest) [inline Mathlib/Algebra/BigOperators/Finprod.lean:359]:
```suggestion
```

File: Mathlib/Algebra/BigOperators/Finprod.lean (hunk 7 above)
Comment (ocfnash) [reply Mathlib/Algebra/BigOperators/Finprod.lean]: Oops I forgot to provide arguments `b 1` to `eq_or_ne`!

File: Mathlib/Algebra/BigOperators/Finprod.lean [hunk 9]
```diff
@@ -351,6 +351,18 @@ theorem finprod_inv_distrib [DivisionCommMonoid G] (f : α → G) : (∏ᶠ x, (
 #align finprod_inv_distrib finprod_inv_distrib
 #align finsum_neg_distrib finsum_neg_distrib
 
+@[to_additive (attr := simp)]
+theorem finprod_prop (P : Prop) [Decidable P] (b : M) : ∏ᶠ _x : P, b = if P then b else 1 := by
+  rw [finprod, dif_pos (toFinite _), Finset.prod_const, mulSupport]
+  simp only [comp_apply, ne_eq]
+  rcases eq_or_ne with rfl | hb; simp
```
Comment (alexjbest) [inline Mathlib/Algebra/BigOperators/Finprod.lean:358]:
```suggestion
  rcases eq_or_ne b 1 with rfl | hb; simp
```

**Comment by eric-wieser** [comment]
Isn't this just `finprod_eq_dif`?

**Comment by eric-wieser** [comment]
bors r-

It is!

File: Mathlib/Algebra/BigOperators/Finprod.lean [hunk 10]
```diff
@@ -351,6 +351,18 @@ theorem finprod_inv_distrib [DivisionCommMonoid G] (f : α → G) : (∏ᶠ x, (
...
+theorem finprod_prop (P : Prop) [Decidable P] (b : M) : ∏ᶠ _x : P, b = if P then b else 1 := by
+  rw [finprod, dif_pos (toFinite _), Finset.prod_const, mulSupport]
+  simp only [comp_apply, ne_eq]
+  rcases eq_or_ne b 1 with rfl | hb; simp
+  · simp only [hb, not_false_eq_true, setOf_true, toFinite_toFinset, toFinset_univ]
+    split_ifs with h
+    · have : Unique (P) := uniqueProp h
+      simp
+    · have : IsEmpty P := ⟨h⟩
+      simp
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Finprod.lean:364]:
```suggestion
theorem finprod_prop (P : Prop) [Decidable P] (b : M) : ∏ᶠ _x : P, b = if P then b else 1 :=
  finprod_eq_if
```
works, so this lemma is a duplicate.

Comment (eric-wieser) [reply Mathlib/Algebra/BigOperators/Finprod.lean:364]: Though I think that's a bad idea too, because then `∏ᶠ x < 37, f x` becomes some awful if.

Comment (alexjbest) [reply Mathlib/Algebra/BigOperators/Finprod.lean:364]: Awful how? I think thats essentially what you want if you have a goal with a finprod over a prop, I see no other way of clearing it tbh.
Or is your example using implicit binders over x also?

Comment (alexjbest) [reply Mathlib/Algebra/BigOperators/Finprod.lean:364]: Also thanks, idk how I missed that lol

Comment (ocfnash) [reply Mathlib/Algebra/BigOperators/Finprod.lean:364]: Oops, I should have spotted this too!

Comment (eric-wieser) [reply Mathlib/Algebra/BigOperators/Finprod.lean:364]: > Or is your example using implicit binders over x also?

Yes, exactly.

Comment (alexjbest) [reply Mathlib/Algebra/BigOperators/Finprod.lean:364]: Yeah this is awkward, I really would like something that applies automatically with simp and norm mum in such a situation, but maybe we just have to wait for better big operator support in norm num

Comment (alexjbest) [reply Mathlib/Algebra/BigOperators/Finprod.lean:364]: Either way I'll close this for now. Sorry for the bother

**Pull request closed without merging**